https://github.com/ultralytics/ultralytics

In [ ]:
#pip install opencv-python

In [ ]:
# 3-5 dakika sürebilir

In [ ]:
#pip install ultralytics           

YOLO, görüntüyü bir kez işleyerek nesneleri tespit etme yeteneğine sahiptir ("You Only Look Once" yani "Bir Kez Bakıyoruz" anlamına gelir).

Görüntü, önceden tanımlanmış bir ölçeğe sahip bir ızgara (grid) bölgesine ayrılır. Her ızgara hücresi belirli bir sayıda tahmin kutusu (bounding box) içerir.

Her bir tahmin kutusu, o bölgede bulunabilecek nesnelerin sınıfını ve konumunu tahmin etmek için kullanılır. Bu tahminler, derin bir evrişimli sinir ağı (CNN) aracılığıyla yapılır.

Sonuç olarak, YOLO aynı anda birden fazla nesneyi tespit edebilir ve bu tespitleri bir çıktı olarak verir.

Özetle, YOLO'nun özel çalışma mantığı, görüntüyü ızgaralara bölmek ve her bir ızgara hücresindeki tahmin kutuları aracılığıyla nesne tespiti yapmaktır. Diğer yandan, CNN genel olarak bir görüntü veri setinden özellikler çıkarmak ve bu özellikleri kullanarak çeşitli görevler gerçekleştirmek için kullanılır. YOLO, içinde CNN tabanlı katmanlar barındıran bir nesne tespiti modelidir.

In [1]:
from ultralytics import YOLO
#from ultralytics.yolo.v8.detect.predict import DetectionPredictor   
from ultralytics.models.yolo.detect import DetectionPredictor    # hata alanlar bu işlemi deneyebilir
import cv2

import numpy as np
import matplotlib.pyplot as plt

**Not:** kamerayı kapatmak için kerneli kapat (shut down)

In [ ]:
# Load a model
model = YOLO("yolov8n.pt")            # load a pretrained model (recommended for training)

# Use the model
results= model.predict(source="0", show=True)
print(results)

In [2]:
model = YOLO("yolov8n.pt")            # load a pretrained model (recommended for training)
result_1 = model.predict("dog.png")


image 1/1 c:\Users\Yaramis\Documents\GitHub\DataScience\Deep Learning\yolo\dog.png: 448x640 1 dog, 726.0ms
Speed: 11.0ms preprocess, 726.0ms inference, 5935.0ms postprocess per image at shape (1, 3, 448, 640)


In [3]:
model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)
result_2 = model.predict("sky.jpg")


image 1/1 c:\Users\Yaramis\Documents\GitHub\DataScience\Deep Learning\yolo\sky.jpg: 384x640 1 airplane, 458.0ms
Speed: 10.0ms preprocess, 458.0ms inference, 8.0ms postprocess per image at shape (1, 3, 384, 640)


In [4]:
model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)
results = model.predict(source="smooth_traffic.mp4")  



WARNING  inference results will accumulate in RAM unless `stream=True` is passed, causing potential out-of-memory
errors for large sources or long-running streams and videos. See https://docs.ultralytics.com/modes/predict/ for help.

Example:
    results = model(source=..., stream=True)  # generator of Results objects
    for r in results:
        boxes = r.boxes  # Boxes object for bbox outputs
        masks = r.masks  # Masks object for segment masks outputs
        probs = r.probs  # Class probabilities for classification outputs

video 1/1 (frame 1/166) c:\Users\Yaramis\Documents\GitHub\DataScience\Deep Learning\yolo\smooth_traffic.mp4: 640x384 9 cars, 1 bus, 1 truck, 404.0ms
video 1/1 (frame 2/166) c:\Users\Yaramis\Documents\GitHub\DataScience\Deep Learning\yolo\smooth_traffic.mp4: 640x384 9 cars, 2 buss, 1 truck, 438.0ms
video 1/1 (frame 3/166) c:\Users\Yaramis\Documents\GitHub\DataScience\Deep Learning\yolo\smooth_traffic.mp4: 640x384 1 person, 9 cars, 2 buss, 523.0ms
video 1/

Bir video dosyasındaki kare sayısının kaç saniyeye denk geldiği, videonun çerçeve hızına (FPS) bağlıdır. FPS, bir videodaki her saniyede görüntülenen kare sayısını ifade eder. Örneğin, bir video 30 FPS ise, her saniyede 30 kare görüntülenir.

Çerçeve hızı bilgisi olmadan doğrudan 166 karenin kaç saniyeye denk geldiğini söylemek zor olur. Ancak genellikle videolar için standart bir çerçeve hızı bulunur. Örneğin, yaygın olarak kullanılan bir çerçeve hızı olan 30 FPS ile hesap yaparsak:

166 / 30 = 5.53 sn 

### mAP@50:95 ile mAP@50 arasında ki fark

mAP@50:95 genellikle daha geniş bir aralığı kapsar ve modelin daha geniş bir doğruluk yelpazesini değerlendirmesine izin verir. Özellikle, daha düşük hassasiyet gerektiren görevlerde (örneğin, geniş nesne tespiti) mAP@50:95 tercih edilebilir. Ancak, bazı durumlarda (örneğin, yüksek hassasiyet gerektiren görevlerde), mAP@50 daha spesifik bir değerlendirme sağlayabilir.